Mathlib/NumberTheory/Cyclotomic/PrimitiveRoots.lean

```diff
@@ -197,6 +217,68 @@ section Norm
 
 namespace IsPrimitiveRoot
 
+section Field
+
+variable {K} [Field K] [NumberField K]
+
+variable (n) in
+/-- If a `n`-th cyclotomic extension of `ℚ` contains a primitive `l`-th root of unity, then
+`l ∣ 2 * n`. -/
+theorem dvd_of_isCyclotomicExtension [NumberField K] [IsCyclotomicExtension {n} ℚ K] {ζ : K}
+    {l : ℕ} (hζ : IsPrimitiveRoot ζ l) (hl : l ≠ 0) : l ∣ 2 * n := by
+  have hl : NeZero l := ⟨hl⟩
+  have hroot := IsCyclotomicExtension.zeta_spec n ℚ K
+  have key := IsPrimitiveRoot.lcm_totient_le_finrank hζ hroot
+    (cyclotomic.irreducible_rat <| Nat.lcm_pos (Nat.pos_of_ne_zero hl.1) n.2)
+  rw [IsCyclotomicExtension.finrank K (cyclotomic.irreducible_rat n.2)] at key
+  rcases _root_.dvd_lcm_right l n with ⟨r, hr⟩
+  have ineq := Nat.totient_super_multiplicative n r
+  rw [← hr] at ineq
+  replace key := (mul_le_iff_le_one_right (Nat.totient_pos n.2)).mp (le_trans ineq key)
+  have rpos : 0 < r := by
+    refine Nat.pos_of_ne_zero (fun h ↦ ?_)
+    simp only [h, mul_zero, _root_.lcm_eq_zero_iff, PNat.ne_zero, or_false] at hr
+    exact hl.1 hr
+  replace key := (Nat.dvd_prime Nat.prime_two).1 (Nat.dvd_two_of_totient_le_one rpos key)
+  rcases key with (key | key)
+  · rw [key, mul_one] at hr
+    rw [← hr]
+    exact dvd_mul_of_dvd_right (_root_.dvd_lcm_left l ↑n) 2
+  · rw [key, mul_comm] at hr
+    simpa [← hr] using _root_.dvd_lcm_left _ _
+
+/-- If `x` is a `k`-th root of unity in an `n`-th cyclotomic extension of `ℚ`, where `n` is odd,
+and `ζ` is a primitive `n`-th root of unity, then there exist `r` such that `x = (-1)^r * ζ^r`. -/
+theorem exists_neg_pow_of_pow_eq_one [NumberField K] [IsCyclotomicExtension {n} ℚ K]
```

Comment:
Isn't there a predicate for being of finite order? or for being a root of unity?

Reply:
`orderOf x \ne 0` should work, but that's not very nice

Reply:
We have `isOfFinOrder`, I've modified the statement to use it.